README

================
What is avrtest?
================

avrtest is a free software simulator for the Atmel AVR family of
microcontrollers distributed under the GNU General Public License.

The main intention of avrtest is to supply a fast, easy-to-use simulator
to run the GCC testsuite for avr-gcc.  Thus, its main audience are guys
and ladies that develop parts of the AVR toolchain like avr-gcc or avr-libc.

avrtest is an instruction set simulator for AVR core families
    avr51: ATmega128*, AT90can128, AT90usb128*
    avr6:  ATmega256*

avrtest-xmega is an instruction set simulator for AVR XMEGA core families
    avrxmega6: ATXmega128*, ...

The executable that supports AVR XMEGA cores is named avrtest-xmega.
In addition to the avrtest simulator, it supports the XMEGA instructions
XCH, LAS, LAC and LAT.

The executable that supports reduced AVR TINY cores is named avrtest-tiny.
It only supports general purpose registers (GPRs) R16..R31,
and many instructions like ADIW are not supported.  The LDS and STS
instructions are 1-word instructions that can access SRAM in the
range 0x40..0xbf.

In the remainder, avrtest is explained.  avrtest-xmega works similar.
avrtest does not simulate internal peripherals like timers, I/O ports,
interrupts, etc.


================
Special features
================

The simulated program can communicate with the simulator: It can write to
the host's standard output and it can read from the host's standard input.
For example, an easy hello world program for AVR could look like this:


#include <stdio.h>

int main (void)
{
    printf ("Hallo World!\n");
    return 0;
}


Compile this program as usual but link against exit-atmega128.o.
How to build that object is explained in the next chapter.  Store the
C source as hello.c and compile it with, e.g.

    avr-gcc hello.c -O -mmcu=atmega128 -o hello.elf /someplace/avrtest/exit-atmega128.o

The exit-*.o object implements stdout as a stream writing through to the
host computer.  Running the program

    avrtest hello.elf

will print something like


Hallo World!

 exit status: EXIT
      reason: exit 0 function called
     program: hello.elf
exit address: 0001a4
total cycles: 837
total instr.: 503


You can also use the respective primitive functions directly:


#include "avrtest.h"

int main (void)
{
    char c;
    
    avrtest_putchar ('Q');

    c = avrtest_getchar();
    avrtest_putchar (c + 1);
    
    return 0;
}


The avrtest_putchar and avrtest_getchar functions are defined in avrtest.h
and act like putchar resp. getchar in an ordinary, hosted program:
putchar writes a character to the host's standard output and getchar reads
a character from the host's standard input.

Compile the C source with

    avr-gcc inout.c -O -mmcu=atmega128 -o inout.elf -I /someplace/avrtest

When running the program with

    avrtest inout.elf

the simulator will print a 'Q' on the console and wait for input.
Type '1<Enter>' and the output will be


Q1
2
 exit status: EXIT
      reason: infinite loop detected (normal exit)
     program: inout.elf
exit address: 0000be
total cycles: 27
total instr.: 18


There are more functions available for program <-> simulator interaction,
see the respective sections below.


=================================================
Running avr-gcc testsuite using avrtest simulator
=================================================

http://lists.gnu.org/archive/html/avr-gcc-list/2009-09/msg00016.html

* Get avr-gcc, Binutils and avr-libc built from sources and working.

* Install DejaGNU, expect and TCL

* Unpack the avrtest tar file from WinAVR sources top "someplace".
  You find a link to the package at "Download Snapshot" on
  https://sourceforge.net/p/winavr/code/HEAD/tree/trunk/avrtest/

* Run

    make

  inside /someplace/avrtest which will build executables from the avrtest
  C-sources and AVR object files exit-*.o from dejagnuboards/exit.c.

* To build exit*.o modules, avr-gcc will be used.  If you want a different
  avr-gcc not found in PATH, use:

    make clean-exit all CC_FOR_AVR=/some-compiler-path/avr-gcc

* Copy /someplace/avrtest/.dejagnurc to your $HOME and change the path
  inside to point to directory where "board file" is located.
  This is dejagnuboards subdirectory of avrtest files:

    set board dir /someplace/avrtest/dejagnuboards

* Edit /someplace/avrtest/dejagnuboards/atmega128-sim.exp and replace paths
  to point to correct location of files for simulator (avrtest executable
  and exit-atmega128.o object file) or add avrtest to $PATH:

    set avrtest_dir "/someplace/avrtest"

  atmega128-sim.exp just sets some custom variables and then loads
  avrtest.exp from the same directory.

* Run the tests from gcc subdirectory of your avr-gcc build directory like:

    make -k check-gcc RUNTESTFLAGS="--target_board=atmega128-sim"
    make -k check-gcc RUNTESTFLAGS="--target_board=atxmega128-sim"

* Look in the ./gcc/testsuite subdirectory of avr-gcc build directory for
  logged results or add the -all switch to see it working 1 test at a time.

    make -k check RUNTESTFLAGS="--target_board=atmega128-sim -all"

* You can run parts of the testsuite like this:

    make -k check RUNTESTFLAGS="--target_board=atmega128-sim avr.exp"
    make -k check RUNTESTFLAGS="--target_board=atmega128-sim avr-torture.exp=pr418854.c"

Voila!


=============================================
 -q : Quiet operation
=============================================

With -q turned on avrtest will print no messages except the ones
explicitly requested by:

  - -runtime
  - LOG_U8 (X), PERF_DUMP_ALL and similar logging commands, see below.

The following exit stati will be returned with -q:

  -  0  Everything went fine.
  -  1  Target program called abort()
  -  x  Target program called exit (x)
  - 10  Program timeout as set by -m MAXCOUNT.
  - 11  Something is wrong with the program file:  Does not fit into
        memory, not an AVR executable, ELF headers broken, ...

  - 20  Out of memory (avrtest_log only).
  - 21  Wrong avrtest usage: Unknown options, etc.
  - 22  Program file could not be found / read.

  - 42  Fatal error in avrtest.


=============================================
 -args ... : passing arguments to the program
=============================================

All arguments after -args will be passed to the target program as if the
program was running from the command line.  This is accomplished by startup
code from exit.c located in section .init8 right before main is called.

If you have own startup code in .init8 make sure it will be located before
the code from exit.c, e.g. by appropriate order of the objects passed to
the linker or by means of a custom linker script.

The first argument (argv[0]) is the program name with directories
stripped off to save space.  The last argument argv[argc] is set to NULL.

The simulator will transfer the arguments to RAM and set R24 to argc,
R22 to argv and R20 to env so that you can write


int main (int argc, char *argv[])
{
    ...
}


resp.


int main (int argc, char *argv[], char *env[])
{
    ...
}


and use argc and argv as in any hosted application:  If the program under
simulation is main.elf and you run

    avrtest main.elf -args hallo "4 2"

then:

    argc       is 3
    argv[0]    is (the address of) string "main.elf" (i.e. argv[0][0] = 'm')
    argv[1]    is (the address of) string "hallo"
    argv[2]    is (the address of) string "4 2"
    argv[3]    is the NULL pointer

env is set to the NULL pointer if the program is executed by avrtest
and to a non-NULL pointer when executed by avrtest_log.

exit.c:avrtest_init_argc_argv() requests the command args to be put
starting at RAM address 0xf000.  If you prefer a different location then
simply adjust exit.c according to your needs following the source comments.


============================
-no-args
============================

will act as if no -args switch was supplied:  It skip all arguments
thereafter and sets argc = 0, argv = NULL and env as described above.


============================
 -no-log and logging control
============================

* This feature is only supported by avrtest_log and avrtest-xmega_log.

avrtest_log will log address and action of each executed instruction to
standard output of the host.  In cases where that is too much a flood of
information, you can start avrtest_log with -no-log and turn on instruction
logging by executing special commands defined in avrtest.h:

    LOG_ON;

Turn on instruction logging.

    LOG_OFF;

Turn it off again.

    LOG_SET (N);

Turn on logging for the next N instructions and then switch it off again.

    LOG_PERF;


NEWS
                                 

                          avrtest NEWS
                          ============

* Support the reduced AVR Tiny cores with only 16 GPRs.    r186  2016-07-11
  The new executables for that core architecture are
  avrtest-tiny and avrtest-tiny_log.

* BUGFIX of wrong decoding of first instruction if         r181  2016-06-25
  entry-point is not 0x0.


* Support call graph generation                            r178  2014-11-13
  - -graph[=FILE]  New option to generate .dot graph


* BUGFIX of wrong SBRC / SBRS / SBIC / SBIS / CPSE         r165  2014-10-20
  skip offset introduced with SVN r157 (2014-10-11).


* The magic ports to establish program <-> simulator       r163  2014-10-19
  interaction have been replaced by the more efficient
  SYSCALLs.


* Show function names from ELF symbol table                r153  2014-10-09
  when logging. 

  - -no-syms       New option to switch that off.


* Use entry point from ELF file.                           r152  2014-10-07


* New magic LOG_PORT for avrtest_log:                      r150  2014-10-06
  (ignored by avrtest)  

  - Turn on / off logging by the simulated program.
  - Capture and print performance data.

  See README for details and examples.


* New options for avrtest_log:                             r144  2014-09-30
  (ignored by avrtest)  

  - -no-log        Don't log instructions as they are simulated.

  - -args ...      Pass all following command options to the
                   running program ready to use as 'argc' and
                   'argv' like in any hosted program:

                       int main (int argc, char *argv[])
                       {
                           /* code */
                       }

  - -no-args ...   Ignore all parameters following -no-args.
                   Set argc=0 and argv=NULL as main is entered.


* Improve log output:                                      r135  2014-09-20

  - Print flags like 'CZNVSHTI' insted of hex value when accessing SREG.
  - Print instruction shortcut for CLR, LSL, ROL, TST when applicable.
  - Better log of conditional jumps and instructions dealing with bits.


* New options:                                             r121  2014-09-15

  -runtime         Show runtime statistics of avrtest itself.

  -e ADDRESS       Use byte address ADDRESS instead of the default
                   program entry point to start the simulation.


* Support AVR Xmega                                        r92   2012-06-04

  - New executables 'avrtest-xega' and 'avrtest-xmega_log'.
  - New Xmega instructions XCH, LAC, LAS, LAT.


* New options:                                             r76   2011-06-09

  -no-stdin        Disable wait and magic action when reading STDIN_PORT.

  -no-stdout       Disable magic action when writing to STDOUT_PORT.
                    

* New header "avrtest.h"

  - Applications can #include "avrtest.h" in their programs or compile
    with -include "avrtest.h" to get macros for the magic ports
    EXIT_PORT, ABORT_PORT, STDOUT_PORT and STDIN_PORT.


* Add support for devices with 3-byte PC                   r52   2008-07-22

  -mmcu=ARCH       New command line option to set the architecture.
                   Valid arguments for ARCH are: 'avr51' and 'avr6'.
                   Default is 'avr51'.


* Support ELF executables                                  r37   2008-05-25

* New options:

  -m CYCLES        Execute at most CYCLES instructions.
                   Default is -m 0 (no limit).

  -d               Initialize SRAM before program execution.


trunk/avrtest/avrtest.h
#ifndef AVRTEST_H
#define AVRTEST_H

#define AVRTEST_INVALID_OPCODE 0xffff

enum
  {
    LOG_ADDR_CMD,
    LOG_STR_CMD,  LOG_SET_FMT_ONCE_CMD,  LOG_SET_FMT_CMD, LOG_UNSET_FMT_CMD,
    LOG_PSTR_CMD, LOG_SET_PFMT_ONCE_CMD, LOG_SET_PFMT_CMD,

    LOG_U8_CMD, LOG_U16_CMD, LOG_U24_CMD, LOG_U32_CMD,
    LOG_S8_CMD, LOG_S16_CMD, LOG_S24_CMD, LOG_S32_CMD,
    LOG_X8_CMD, LOG_X16_CMD, LOG_X24_CMD, LOG_X32_CMD,
    LOG_FLOAT_CMD,

    LOG_TAG_FMT_CMD, LOG_TAG_PFMT_CMD,

    LOG_X_sentinel
  };

enum
  {
    TICKS_GET_CYCLES_CMD,
    TICKS_GET_INSNS_CMD,
    TICKS_GET_RAND_CMD, 
    TICKS_GET_PRAND_CMD, 

    TICKS_RESET_CYCLES_CMD = 1 << 2,
    TICKS_RESET_INSNS_CMD  = 1 << 3,
    TICKS_RESET_PRAND_CMD  = 1 << 4,
    TICKS_RESET_ALL_CMD = TICKS_RESET_CYCLES_CMD
                          | TICKS_RESET_INSNS_CMD
                          | TICKS_RESET_PRAND_CMD
  };

enum
  {
    PERF_STOP_CMD, PERF_DUMP_CMD,
    PERF_START_CMD, PERF_START_CALL_CMD,
    PERF_STAT_U32_CMD, PERF_STAT_S32_CMD, PERF_STAT_FLOAT_CMD,

    PERF_STAT_CMD = PERF_STAT_U32_CMD
  };

enum
  {
    PERF_TAG_STR_CMD, PERF_TAG_S32_CMD, PERF_TAG_U32_CMD,
    PERF_TAG_S16_CMD, PERF_TAG_U16_CMD,
    PERF_TAG_FLOAT_CMD,
    PERF_LABEL_CMD, PERF_PLABEL_CMD,
    PERF_TAG_FMT_CMD, PERF_TAG_PFMT_CMD
  };

#ifdef IN_AVRTEST

/* This defines are for avrtest itself.  */

/* bits 5..3 */
#define PERF_CMD(n) (0xf & ((n) >> 4))

#define PERF_TAG_CMD(n) (0xf & ((n) >> 4))

/* bits 2..0 (LOG_CMD = 0) */
/* bits 2..0 (LOG_CMD = LOG_TAG_PERF) */
#define PERF_N(n)   ((n) & 0x7)

/* mask */
#define PERF_ALL    0xfe

#else /* IN_AVRTEST */

/* This defines can be used in the AVR application.  */

/* Logging Control */
#define LOG_TAG_CMD(N,T) ((((PERF_## T ##_CMD) & 0xf) << 4) | ((N) & 7))

/* Performance Measurement */
#define PERF_CMD_(n,c)       (((n) & 7) | (PERF_## c ##_CMD << 4))

/* Convenience */

/* Logging with default format strings */
#define LOG_ADDR(X)  avrtest_syscall_7_a ((X), LOG_ADDR_CMD)
#define LOG_PSTR(X)  avrtest_syscall_7_a ((X), LOG_PSTR_CMD)
#define LOG_STR(X)   avrtest_syscall_7_a ((X), LOG_STR_CMD)
#define LOG_FLOAT(X) avrtest_syscall_7_f ((X), LOG_FLOAT_CMD)

#define LOG_U8(X)  avrtest_syscall_7_u8  ((X), LOG_U8_CMD)
#define LOG_U16(X) avrtest_syscall_7_u16 ((X), LOG_U16_CMD)
#define LOG_U24(X) avrtest_syscall_7_u24 ((X), LOG_U24_CMD)
#define LOG_U32(X) avrtest_syscall_7_u32 ((X), LOG_U32_CMD)

#define LOG_S8(X)  avrtest_syscall_7_s8  ((X), LOG_S8_CMD)
#define LOG_S16(X) avrtest_syscall_7_s16 ((X), LOG_S16_CMD)
#define LOG_S24(X) avrtest_syscall_7_s24 ((X), LOG_S24_CMD)
#define LOG_S32(X) avrtest_syscall_7_s32 ((X), LOG_S32_CMD)

#define LOG_X8(X)  avrtest_syscall_7_u8  ((X), LOG_X8_CMD)
#define LOG_X16(X) avrtest_syscall_7_u16 ((X), LOG_X16_CMD)
#define LOG_X24(X) avrtest_syscall_7_u24 ((X), LOG_X24_CMD)
#define LOG_X32(X) avrtest_syscall_7_u32 ((X), LOG_X32_CMD)

/* Format string for next action only */
#define LOG_SET_FMT_ONCE(F)  avrtest_syscall_7_a ((F), LOG_SET_FMT_ONCE_CMD)
#define LOG_SET_PFMT_ONCE(F) avrtest_syscall_7_a ((F), LOG_SET_PFMT_ONCE_CMD)

/* Setting custom format string for all subsequent logs */
#define LOG_SET_FMT(F)   avrtest_syscall_7_a ((F), LOG_SET_FMT_CMD)
#define LOG_SET_PFMT(F)  avrtest_syscall_7_a ((F), LOG_SET_PFMT_CMD)

/* Reset to default format strings */
#define LOG_UNSET_FMT(F) avrtest_syscall_7 (LOG_UNSET_FMT_CMD)

/* Logging with custom format string (from RAM) */
#define LOG_FMT_ADDR(F,X)  LOG_DUMP_F_ (FMT, (F), LOG_ADDR(X))
#define LOG_FMT_PSTR(F,X)  LOG_DUMP_F_ (FMT, (F), LOG_PSTR(X))
#define LOG_FMT_STR(F,X)   LOG_DUMP_F_ (FMT, (F), LOG_STR(X))
#define LOG_FMT_FLOAT(F,X) LOG_DUMP_F_ (FMT, (F), LOG_FLOAT(X))

#define LOG_FMT_U8(F,X)  LOG_DUMP_F_ (FMT, (F), LOG_U8(X))
#define LOG_FMT_U16(F,X) LOG_DUMP_F_ (FMT, (F), LOG_U16(X))
#define LOG_FMT_U24(F,X) LOG_DUMP_F_ (FMT, (F), LOG_U24(X))
#define LOG_FMT_U32(F,X) LOG_DUMP_F_ (FMT, (F), LOG_U32(X))

#define LOG_FMT_S8(F,X)  LOG_DUMP_F_ (FMT, (F), LOG_S8(X))
#define LOG_FMT_S16(F,X) LOG_DUMP_F_ (FMT, (F), LOG_S16(X))
#define LOG_FMT_S24(F,X) LOG_DUMP_F_ (FMT, (F), LOG_S24(X))
#define LOG_FMT_S32(F,X) LOG_DUMP_F_ (FMT, (F), LOG_S32(X))

#define LOG_FMT_X8(F,X)  LOG_DUMP_F_ (FMT, (F), LOG_X8(X))
#define LOG_FMT_X16(F,X) LOG_DUMP_F_ (FMT, (F), LOG_X16(X))
#define LOG_FMT_X24(F,X) LOG_DUMP_F_ (FMT, (F), LOG_X24(X))
#define LOG_FMT_X32(F,X) LOG_DUMP_F_ (FMT, (F), LOG_X32(X))

/* Logging with custom format string (from Flash) */
#define LOG_PFMT_ADDR(F,X)  LOG_DUMP_F_ (PFMT, (F), LOG_ADDR(X))
#define LOG_PFMT_PSTR(F,X)  LOG_DUMP_F_ (PFMT, (F), LOG_PSTR(X))
#define LOG_PFMT_STR(F,X)   LOG_DUMP_F_ (PFMT, (F), LOG_STR(X))
#define LOG_PFMT_FLOAT(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_FLOAT(X))

#define LOG_PFMT_U8(F,X)  LOG_DUMP_F_ (PFMT, (F), LOG_U8(X))
#define LOG_PFMT_U16(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_U16(X))
#define LOG_PFMT_U24(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_U24(X))
#define LOG_PFMT_U32(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_U32(X))

#define LOG_PFMT_S8(F,X)  LOG_DUMP_F_ (PFMT, (F), LOG_S8(X))
#define LOG_PFMT_S16(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_S16(X))
#define LOG_PFMT_S24(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_S24(X))
#define LOG_PFMT_S32(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_S32(X))

#define LOG_PFMT_X8(F,X)  LOG_DUMP_F_ (PFMT, (F), LOG_X8(X))
#define LOG_PFMT_X16(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_X16(X))
#define LOG_PFMT_X24(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_X24(X))
#define LOG_PFMT_X32(F,X) LOG_DUMP_F_ (PFMT, (F), LOG_X32(X))

#define LOG_DUMP_F_(W, F, C)                                \
  do {                                                      \
      avrtest_syscall_7_a ((F), LOG_SET_## W ##_ONCE_CMD);  \
      C;                                                    \
    } while (0)

/* Tagging perf-meter rounds */

/* Tagging using default format string */
#define PERF_TAG_STR(N,X)   avrtest_syscall_6_s ((X), LOG_TAG_CMD (N, TAG_STR))
#define PERF_TAG_S16(N,X)   avrtest_syscall_6_i ((X), LOG_TAG_CMD (N, TAG_S16))
#define PERF_TAG_S32(N,X)   avrtest_syscall_6_l ((X), LOG_TAG_CMD (N, TAG_S32))
#define PERF_TAG_U16(N,X)   avrtest_syscall_6_u ((X), LOG_TAG_CMD (N, TAG_U16))
#define PERF_TAG_U32(N,X)   avrtest_syscall_6_ul((X), LOG_TAG_CMD (N, TAG_U32))
#define PERF_TAG_FLOAT(N,X) avrtest_syscall_6_f ((X), LOG_TAG_CMD (N, TAG_FLOAT))

/* Labeling perf-meter */
#define PERF_LABEL(N,L)     avrtest_syscall_6_s ((L), LOG_TAG_CMD (N, LABEL))
#define PERF_PLABEL(N,L)    avrtest_syscall_6_s ((L), LOG_TAG_CMD (N, PLABEL))

/* Set custom format string for tagging */
#define PERF_TAG_FMT(F)     avrtest_syscall_6_s (F, LOG_TAG_CMD (0, TAG_FMT))
#define PERF_TAG_PFMT(F)    avrtest_syscall_6_s (F, LOG_TAG_CMD (0, TAP_PFMT))

/* Tagging with custom format string (from RAM) */
#define PERF_TAG_FMT_STR(N,F,X)   PERF_TAG_F_((F), (X), (N), _6_s,  TAG_STR)
#define PERF_TAG_FMT_S16(N,F,X)   PERF_TAG_F_((F), (X), (N), _6_i,  TAG_S16)
#define PERF_TAG_FMT_S32(N,F,X)   PERF_TAG_F_((F), (X), (N), _6_l,  TAG_S32)
#define PERF_TAG_FMT_U16(N,F,X)   PERF_TAG_F_((F), (X), (N), _6_u,  TAG_U16)
#define PERF_TAG_FMT_U32(N,F,X)   PERF_TAG_F_((F), (X), (N), _6_ul, TAG_U32)
#define PERF_TAG_FMT_FLOAT(N,F,X) PERF_TAG_F_((F), (X), (N), _6_f,  TAG_FLOAT)

/* Tagging with custom format string (from Flash) */
#define PERF_TAG_PFMT_STR(N,F,X)   PERF_TAG_PF_((F), (X), (N), _6_s,  TAG_STR)
#define PERF_TAG_PFMT_S16(N,F,X)   PERF_TAG_PF_((F), (X), (N), _6_i   TAG_S16)
#define PERF_TAG_PFMT_S32(N,F,X)   PERF_TAG_PF_((F), (X), (N), _6_l   TAG_S32)
#define PERF_TAG_PFMT_U16(N,F,X)   PERF_TAG_PF_((F), (X), (N), _6_u,  TAG_U16)
#define PERF_TAG_PFMT_U32(N,F,X)   PERF_TAG_PF_((F), (X), (N), _6_ul, TAG_U32)
#define PERF_TAG_PFMT_FLOAT(N,F,X) PERF_TAG_PF_((F), (X), (N), _6_f,  TAG_FLOAT)

#define PERF_TAG_F_(F,X,N,S,C) \
  do { PERF_TAG_FMT(F); avrtest_syscall##S (X, LOG_TAG_CMD (N,C)); } while (0)

#define PERF_TAG_PF_(F,X,N,S,C) \
  do { PERF_TAG_PFMT(F); avrtest_syscall##S (X, LOG_TAG_CMD (N,C)); } while (0)

/* Instruction logging control */

#define LOG_SET(N)    avrtest_syscall_3 (N)
#define LOG_PERF      avrtest_syscall_2 ()
#define LOG_ON        avrtest_syscall_1 ()
#define LOG_OFF       avrtest_syscall_0 ()

/* Controling perf-meters */
#define PERF_STOP(n)        avrtest_syscall_5 (PERF_CMD_((n), STOP))
#define PERF_START(n)       avrtest_syscall_5 (PERF_CMD_((n), START))
#define PERF_DUMP(n)        avrtest_syscall_5 (PERF_CMD_((n), DUMP))
#define PERF_START_CALL(n)  avrtest_syscall_5 (PERF_CMD_((n), START_CALL))

#define PERF_STOP_ALL         PERF_STOP (0)
#define PERF_START_ALL        PERF_START (0)
#define PERF_START_CALL_ALL   PERF_START_CALL (0)
#define PERF_DUMP_ALL         PERF_DUMP (0)

/* Perf-meter Min/Max on value from program */
#define PERF_STAT_U32(n,x)   avrtest_syscall_5_u ((x), PERF_CMD_((n),STAT_U32))
#define PERF_STAT_S32(n,x)   avrtest_syscall_5_s ((x), PERF_CMD_((n),STAT_S32))
#define PERF_STAT_FLOAT(n,x) avrtest_syscall_5_f ((x), PERF_CMD_((n),STAT_FLOAT))

#define AT_INLINE __inline__ __attribute__((__always_inline__))

#define CPSE_rr_(r)                         \
  (((0UL + AVRTEST_INVALID_OPCODE) << 16)   \
   | 0x1000 | (r << 4) | (r & 0xf) | ((r & 0x10) << 5))

__extension__ enum
{
  SYSCo_0 = CPSE_rr_(0),  SYSCo_10 = CPSE_rr_(10),  SYSCo_20 = CPSE_rr_(20),
  SYSCo_1 = CPSE_rr_(1),  SYSCo_11 = CPSE_rr_(11),  SYSCo_21 = CPSE_rr_(21),
  SYSCo_2 = CPSE_rr_(2),  SYSCo_12 = CPSE_rr_(12),  SYSCo_22 = CPSE_rr_(22),
  SYSCo_3 = CPSE_rr_(3),  SYSCo_13 = CPSE_rr_(13),  SYSCo_23 = CPSE_rr_(23),
  SYSCo_4 = CPSE_rr_(4),  SYSCo_14 = CPSE_rr_(14),  SYSCo_24 = CPSE_rr_(24),
  SYSCo_5 = CPSE_rr_(5),  SYSCo_15 = CPSE_rr_(15),  SYSCo_25 = CPSE_rr_(25),
  SYSCo_6 = CPSE_rr_(6),  SYSCo_16 = CPSE_rr_(16),  SYSCo_26 = CPSE_rr_(26),
  SYSCo_7 = CPSE_rr_(7),  SYSCo_17 = CPSE_rr_(17),  SYSCo_27 = CPSE_rr_(27),
  SYSCo_8 = CPSE_rr_(8),  SYSCo_18 = CPSE_rr_(18),  SYSCo_28 = CPSE_rr_(28),
  SYSCo_9 = CPSE_rr_(9),  SYSCo_19 = CPSE_rr_(19),  SYSCo_29 = CPSE_rr_(29),

  SYSCo_30 = CPSE_rr_(30),
  SYSCo_31 = CPSE_rr_(31),
};

#define AVRTEST_DEF_SYSCALL0(S, N)                          \
  static AT_INLINE                                          \
  void avrtest_syscall ## S (void)                          \
  {                                                         \
    __asm __volatile__ (".long %1 ;; SYSCALL %0"            \
                        :: "n" (N), "n" (SYSCo_ ## N));     \
  }

#define AVRTEST_DEF_SYSCALL1(S, N, T1, R1)                  \
  static AT_INLINE                                          \
  void avrtest_syscall ## S (T1 _v1_)                       \
  {                                                         \
    register T1 r##R1 __asm (#R1) = _v1_;                   \
    __asm __volatile__ (".long %1 ;; SYSCALL %0"            \
                        :: "n" (N), "n" (SYSCo_ ## N),      \
                        "r" (r##R1));                       \
  }

#define AVRTEST_DEF_SYSCALL2(S, N, T1, R1, T2, R2)          \
  static AT_INLINE                                          \
  void avrtest_syscall ## S (T1 _v1_, T2 _v2_)              \
  {                                                         \
    register T1 r##R1 __asm (#R1) = _v1_;                   \
    register T2 r##R2 __asm (#R2) = _v2_;                   \
    __asm __volatile__ (".long %1 ;; SYSCALL %0"            \
                        :: "n" (N), "n" (SYSCo_ ## N),      \
                        "r" (r##R1), "r" (r##R2));          \
  }

#define AVRTEST_DEF_SYSCALL1_0(S, N, T0, R0)                \
  static AT_INLINE                                          \
  T0 avrtest_syscall ## S (void)                            \
  {                                                         \
    register T0 r##R0 __asm (#R0);                          \
    __asm __volatile__ (".long %2 ;; SYSCALL %1"            \
                        : "=r" (r##R0)                      \
                        : "n" (N), "n" (SYSCo_ ## N));      \
    return r##R0;                                           \
  }

#define AVRTEST_DEF_SYSCALL1_1(S, N, T0, R0, T1, R1)        \
  static AT_INLINE                                          \
  T0 avrtest_syscall ## S (T1 _v1_)                         \
  {                                                         \
    register T0 r##R0 __asm (#R0);                          \
    register T1 r##R1 __asm (#R1) = _v1_;                   \
    __asm __volatile__ (".long %2 ;; SYSCALL %1"            \
                        : "=r" (r##R0)                      \
                        : "n" (N), "n" (SYSCo_ ## N),       \
                          "r" (r##R1));                     \
    return r##R0;                                           \
  }


AVRTEST_DEF_SYSCALL1 (_27, 27, void*, 24)
AVRTEST_DEF_SYSCALL1_0 (_28, 28, int, 24)
AVRTEST_DEF_SYSCALL1 (_29, 29, char, 24)
AVRTEST_DEF_SYSCALL1 (_30, 30, int, 24)
AVRTEST_DEF_SYSCALL0 (_31, 31)

AVRTEST_DEF_SYSCALL0 (_0, 0) /* LOG_OFF  */
AVRTEST_DEF_SYSCALL0 (_1, 1) /* LOG_ON   */
AVRTEST_DEF_SYSCALL0 (_2, 2) /* LOG_PERF */

/* LOG_SET (N) */
AVRTEST_DEF_SYSCALL1 (_3, 3, unsigned, 24)

/* Cycle count, instruction cound, (pseudo) random number */
AVRTEST_DEF_SYSCALL1   (_4_r, 4, unsigned char, 24)
AVRTEST_DEF_SYSCALL1_1 (_4_g, 4, unsigned long, 22, unsigned char, 24)

/* Perf-meter control */
AVRTEST_DEF_SYSCALL1 (_5, 5, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_5_u, 5, unsigned long, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_5_s, 5, signed long,   20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_5_f, 5, float,         20, unsigned char, 24)

/* PERF_TAG */
AVRTEST_DEF_SYSCALL2 (_6_s, 6, const volatile char*, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_6_i, 6, int, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_6_u, 6, unsigned, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_6_l, 6, long, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_6_ul, 6, unsigned long, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_6_f,  6, float, 20, unsigned char, 24)

/* Logging values */
AVRTEST_DEF_SYSCALL1 (_7, 7, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_a, 7, const volatile void*, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_f, 7, float,                20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_u8,  7, unsigned char, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_s8,  7,   signed char, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_u16, 7, unsigned int,  20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_s16, 7,   signed int,  20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_u32, 7, unsigned long, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_s32, 7,   signed long, 20, unsigned char, 24)

#ifdef __UINT24_MAX__
AVRTEST_DEF_SYSCALL2 (_7_u24, 7, __uint24, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_s24, 7, __int24,  20, unsigned char, 24)
#else
AVRTEST_DEF_SYSCALL2 (_7_u24, 7, unsigned long, 20, unsigned char, 24)
AVRTEST_DEF_SYSCALL2 (_7_s24, 7,   signed long, 20, unsigned char, 24)
#endif


#undef AVRTEST_DEF_SYSCALL0
#undef AVRTEST_DEF_SYSCALL1
#undef AVRTEST_DEF_SYSCALL2
#undef AVRTEST_DEF_SYSCALL1_0
#undef AVRTEST_DEF_SYSCALL1_1

static AT_INLINE void
avrtest_abort (void)
{
  avrtest_syscall_31 ();
}

static AT_INLINE void
avrtest_exit (int _status)
{
  avrtest_syscall_30 (_status);
}

static AT_INLINE void
avrtest_putchar (int _c)
{
  avrtest_syscall_29 (_c);
}

static AT_INLINE int
avrtest_getchar (void)
{
  return avrtest_syscall_28 ();
}

static AT_INLINE unsigned long
avrtest_cycles (void)
{
  return avrtest_syscall_4_g (TICKS_GET_CYCLES_CMD);
}

static AT_INLINE unsigned long
avrtest_insns (void)
{
  return avrtest_syscall_4_g (TICKS_GET_INSNS_CMD);
}

static AT_INLINE unsigned long
avrtest_rand (void)
{
  return avrtest_syscall_4_g (TICKS_GET_RAND_CMD);
}

static AT_INLINE unsigned long
avrtest_prand (void)
{
  return avrtest_syscall_4_g (TICKS_GET_PRAND_CMD);
}

static AT_INLINE void
avrtest_reset_cycles (void)
{
  avrtest_syscall_4_r (TICKS_RESET_CYCLES_CMD);
}

static AT_INLINE void
avrtest_reset_insns (void)
{
  avrtest_syscall_4_r (TICKS_RESET_INSNS_CMD);
}

static AT_INLINE void
avrtest_reset_prand (void)
{
  avrtest_syscall_4_r (TICKS_RESET_PRAND_CMD);
}

static AT_INLINE void
avrtest_reset_all (void)
{
  avrtest_syscall_4_r (TICKS_RESET_ALL_CMD);
}

#undef AT_INLINE

#endif /* IN_AVRTEST */

#endif /* AVRTEST_H */
